### DISCUSSION TITLE
How do I do a Custom Text Editor plugin?

### DISCUSSION THREAD

**Discussion by IvanFarkas** [facebook/lexical#3700] [Q&A]
# How do I do a Custom SSML (text) Editor plugin?

**The documentation is very weak when it comes to the Custom Plugin.**

## I'd like to do an [SSML](https://www.w3.org/TR/speech-synthesis11/) Editor in Next.js, with TypeScript.
Basic functionality:

- Selects text
- Click the toolbar icon, like Sentence
- Editor wraps selection in SSML elements, like `<s></s>`.
- If the mark has attributes a popup is provided to enter. just like Link in RTE.

I am already using the RTE in my code successfully and understand the concepts.
I looked at the following: 

- [Emoticons](https://github.com/facebook/lexical/blob/main/examples/emoticons.md)
- [@lexical/react](https://github.com/facebook/lexical/tree/main/packages/lexical-react)
- [@lexical/mark](https://github.com/facebook/lexical/tree/main/packages/lexical-mark)
- [@lexical/plain-text](https://github.com/facebook/lexical/tree/main/packages/lexical-plain-text)
- [lexical-playground plugins](https://github.com/facebook/lexical/tree/main/packages/lexical-playground/src/plugins)

It's still unclear to me what's the best, clean way to do this, although I am leaning toward somehow extending the @lexical/react.
Any help or pointers are greatly appreciated.
Docs need work.

**Comment by IvanFarkas** [comment]
Perhaps Extending ElementNode?

**Reply by svssrinivas** [reply]
@IvanFarkas I am experimenting with something similar. I have extended `ElementNode` to create nodes for custom XML tags. My requirements to add XML tags are similar to yours.

When I try to insert a custom `ElementNode` at selection, the contents of selection are lost. Is there some way to add the contents of`RangeSelection` back to the newly created node? 

Additionally, I am not sure of how to handle the case where the selection starts / ends in the middle of a `TextNode`.
